keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+  behavior there.
+
+[Proxy Terminating Endpoints]: ../1669-graceful-termination-local-external-traffic-policy/
+
+## Proposal
+
+### User Stories
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
+topology and zone-level topology, which is not necessarily terrible;
+the zone-level topology would only be useful when the node-level
+topology "failed". But if we wanted to cover that case, then we could
+have a separate annotation for node-level topology.)
+
+#### Approach 2: Automatic
+
+Alternatively, the EndpointSlice controller could attempt to determine
+automatically which services would benefit from node-level topology.
+For example, the EndpointSlice controller could look for services
+where:
+
+  - The service has an endpoint on every node, or at least "almost
+    every" node. (eg, no more than N or N% of nodes are missing an
+    endpoint).
```

Comment:
On some occasions I've seen clusters where there are multiple fairly isolated node pools, where a daemonset or set of Pods runs exclusively on a specific node pool. That kind of situation seems like it could benefit from this PreferLocal kind of approach, but this automatic approach does not appear to be compatible. Is there any merit in an approach that simply routes traffic to a local endpoint if it exists and otherwise falls back to default routing?

That's actually already how zonal topology hints work (https://github.com/kubernetes/kubernetes/blob/3af1e5fdf6f3d3203283950c1c501739c21a53e2/pkg/proxy/topology.go#L178).

Reply:
We can't automatically recognize that node-level topology would be useful in cases like that, though, because just from looking at the Service/DaemonSet, you can't tell how the _clients_ of that service are going to be distributed, so you can't tell if node-level topology would help or hurt.

So that's an argument in favor of the manual approach.

Reply:
What if the criteria is more like:

For those nodes which DO have at least 1 endpoint, if they all have approximately the same number of endpoints (maybe some %age threshold so we don't end up saying "1 is approximately 2") then prefer-local is allowed.  

or even:

Any node with at least the mean number of endpoints can use a prefer-local strategy, and any node with less than the mean must fall back on something else.

Does that work?  Does it make this mode more amenable to being automatic?